services/storage/src/simcore_service_storage/simcore_s3_dsm.py

```diff
@@ -886,17 +905,46 @@ async def _copy_file_s3_s3(
                 user_id,
                 dst_file_id,
                 upload_id=S3_UNDEFINED_OR_EXTERNAL_MULTIPART_ID,
-                is_directory=False,
+                is_directory=src_fmd.is_directory,
             )
             # NOTE: ensure the database is updated so cleaner does not pickup newly created uploads
             await transaction.commit()
 
-            await get_s3_client(self.app).copy_file(
-                self.simcore_bucket_name,
-                src_fmd.object_name,
-                new_fmd.object_name,
-                bytes_transfered_cb=bytes_transfered_cb,
-            )
+            s3_client: StorageS3Client = get_s3_client(self.app)
+
+            if src_fmd.is_directory:
+                async for s3_objects in s3_client.list_all_objects_iter(
+                    self.simcore_bucket_name,
+                    prefix=src_fmd.object_name,
+                    max_yield_result_size=_MAX_ELEMENTS_TO_LIST,
+                ):
+                    s3_objects_src_to_new: dict[str, str] = {
+                        x["Key"]: x["Key"].replace(
+                            f"{src_fmd.object_name}", f"{new_fmd.object_name}"
+                        )
+                        for x in s3_objects
+                    }
+
+                    await logged_gather(
+                        *[
+                            s3_client.copy_file(
+                                self.simcore_bucket_name,
+                                parse_obj_as(SimcoreS3FileID, src),
```

Comment:
are you sure the validation is needed here? just casting is enough no?

Reply:
`parse_obj_as(SimcoreS3FileID, src)`, here `src` is coming from S3 API so a cast can be enough.
`parse_obj_as(SimcoreS3FileID, new)` this one is composed on the fly and it might be better to always validate.

Reply:
so one will have a cast and one a validator

Reply:
check if the validation is really needed. over-validating in vase of 10000 files will be costly

Reply:
I see the point now.
So we can do without the validation. We are replacing the `{project_id}/{node_id}` with one coming from a validated object. So as you pointed out, there is no need to validate it.
Left only the cast.

Reply:
@sanderegg